### PR TITLE
Tracking tests

### PR DESCRIPTION
**Pull request by gferraro**
Adds some scripts to create some classyf/tracking tests on existing clips
Then test that we are getting similar results

Also added downloading a model and classifying  to travis

### CODE REVIEW

File: tests/tracking-tests.yml [hunk 1]
```diff
@@ -0,0 +1,362 @@
+!TestConfig
+clip_dir: tests/test-clips
+recording_tests:
+- !TestRecording
+  device: Wallaby
```
Comment (clare) [inline tests/tracking-tests.yml:5]: I think it is really interesting to know why this file was chosen.   Especially for people who are looking at the code later.  For example: 'Animal already in frame' or 'very small (inditinguishable) blob' or 'blurry insect'.   

Thus if the test fails it makes it far more motivating to fix it!

Mabye So have a summary: tag.

File: tests/tracking-tests.yml [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+  device_id: 956
+  filename: 551732.cptv
+  group: ORC
+  group_id: 160
+  rec_id: 551732
+  tracks:
+  - !TestTrack
+    confidence: 0.85
```
Comment (clare) [inline tests/tracking-tests.yml:13]: Is start and end frame enough or do we need the full tracks?  (In which case we would definitely need some way of seeing the different tracking and and automatic way of diffing to see whether changes are significant or insignificant.

File: tests/tracking-tests.yml [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    opt_start: 3.0
+    start: 0
+    start_pos:
+    - 0
+    - - 4
+      - 91
+      - 14
+      - 101
+    tag: bird
+- !TestRecording
```
Comment (clare) [inline tests/tracking-tests.yml:72]: Would probably be more manageable if each recording was independant in a separate file.   Then if you add a new test no 'old tests files' will need to change.

File: tests/trackingtest.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+def init_logging():
+    logging.root.removeHandler(absl.logging._absl_handler)
+    absl.logging._warn_preinit_stderr = False
+    logging.basicConfig(stream=sys.stderr, level=logging.INFO)
+
+
+def main():
+    init_logging()
+    args = parse_args()
+    test = TestClassify(args)
```
Comment (clare) [inline tests/trackingtest.py:393]: Are these tracking tests or classify tests?

File: tests/trackingtest.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+            keras_model,
+        )
+        # try download missing tests
+        if args.user and args.password:
+            api = API(args.user, args.password, args.server)
+            out_dir = Path(self.test_config.clip_dir)
+            if not out_dir.exists():
+                out_dir.mkdir()
+            for test in self.test_config.recording_tests:
+                filepath = out_dir / test.filename
```
Comment (clare) [inline tests/trackingtest.py:257]: Can this code be moved into the API - there is similar code in generate tests.

File: tests/trackingtest.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+        expected_length = test_track.opt_end - test_track.opt_start
+        self.length_diff = round(expected_length - (track.end_s - track.start_s), 2)
+        self.start_diff_s = round(test_track.start - track.start_s, 2)
+        self.end_diff_s = round(test_track.end - track.end_s, 2)
+        self.opt_start_diff_s = round(test_track.opt_start - track.start_s, 2)
+        self.opt_end_diff_s = round(test_track.opt_end - track.end_s, 2)
+        self.error = round(abs(self.opt_start_diff_s) + abs(self.opt_end_diff_s), 1)
+
+        if self.error <= test_track.calc_error():
+            self.status = 1
```
Comment (clare) [inline tests/trackingtest.py:178]: Status should be an enum not a **magic number** - 1.  More information https://docs.python.org/3/library/enum.html

File: tests/trackingtest.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+            self.status = 0
+        else:
+            self.status = -1
+        self.expected_tag = test_track.tag
+        self.got_animal = tag
+        self.test_track = test_track
+        self.track = track
+
+    def tracking_status(self):
+        if self.status == 1:
```
Comment (clare) [inline tests/trackingtest.py:189]: If you used a enum you could probably get away with just using enum.name and wouldn't need this function.

File: tests/trackingtest.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+MATCH_ERROR = 1
+
+
+def match_track(gen_track, expected_tracks):
+    score = None
+    match = None
+    MAX_ERROR = 8
+    for track in expected_tracks:
+        start_diff = abs(track.start - gen_track.start_s)
```
Comment (clare) [inline tests/trackingtest.py:23]: This is clever - but is it really needed?   Can we just order the tracks by time and the compare the first with the first, etc.   If there is a different number of tracks chances are we will need to manually inspect the tracking to see if it ok.   

The key here is the normally we would expect the computed tracks to be very very similar to the saved tracks.   If they aren't we probably need to compare the two tracks to see which is right.

File: tests/trackingtest.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+        if total_summary is None:
+            return
+        classified_correct = total_summary.get("classify", {}).get("correct", 0)
+        tracked_well = total_summary.get("tracking", {}).get("better", 0)
+        tracked_well += total_summary.get("tracking", {}).get("same", 0)
+        unmatched_tracks = total_summary.get("tracking", {}).get("unmatched_tracks", 0)
+        unmatched_tests = total_summary.get("tracking", {}).get("unmatched_tests", 0)
+
+        classified_per = round(100.0 * classified_correct / total_tracks)
+        tracked_per = round(100.0 * tracked_well / total_tracks)
```
Comment (clare) [inline tests/trackingtest.py:320]: What if there are extra tracks or missing tracks.   Is it possible to get 100% tracked well?  

Eg If there is one missing track but the existing tracks are perfect won't it come out as perfect?

File: tests/trackingtest.py (hunk 7 above)
Comment (clare) [reply tests/trackingtest.py:189]: How can you be sure it is better tracking?   Just because the track is longer or shorter?

File: tests/trackingtest.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+        self.opt_end_diff_s = round(test_track.opt_end - track.end_s, 2)
+        self.error = round(abs(self.opt_start_diff_s) + abs(self.opt_end_diff_s), 1)
+
+        if self.error <= test_track.calc_error():
+            self.status = 1
+        elif self.error < MATCH_ERROR:
+            self.status = 0
+        else:
+            self.status = -1
+        self.expected_tag = test_track.tag
```
Comment (clare) [inline tests/trackingtest.py:183]: Are you even comparing the actual bounding boxing in working out if the tracking is the same?

**Review by clare — APPROVED** [review]
This is a good start, but is quite hard to follow exactly what is happening.  

A lot of experience in writing tests has taught me the value of keeping things super simple programming wise as no one want to debug test code.  
Knowing that we are using the same files the tests results should be very very close to our saved results so we should be able to use this to our advantage. 

For now I think you need to:
  - Make status an enum
  - Make sure you can't get a 100% match even if there are extra or missing tracks. 

In the future we should look at:
  - splitting the different cptv files into separate tests.
  - using the pytest framework to run the tests.
  - comparing the bounding boxes the whole way through the track.

File: tests/api.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
+import requests
```
[the remark this answers is not in the document]
Comment (gferraro) [reply tests/api.py:1]: There is it came out after this PR

File: tests/trackingtest.py (hunk 7 above)
Comment (gferraro) [reply tests/trackingtest.py:189]: Ye its a very simple metric, i think of start and end being the "expected tracking" while opt start and end being "the perfect tracking normally unattaible" so if a track is closer to perfect than expect i deem it as improved.

File: tests/trackingtest.py (hunk 8 above)
Comment (gferraro) [reply tests/trackingtest.py:23]: You are probably right, i was doing it just by ordering before but in some cases were getting mis matches. But i am happy to say that these cases should be manually checked

File: tests/trackingtest.py (hunk 10 above)
Comment (gferraro) [reply tests/trackingtest.py:183]: Thats what match_track function does on line 26

File: tests/trackingtest.py (hunk 9 above)
Comment (gferraro) [reply tests/trackingtest.py:320]: I might be missing something here but i dont think that is the case.
If there is 2 total tracks, and 1 perfect track 
then tracked_well will be 1 and total_tracks will be 2  = 50%

It is possible if all the new tracks matched to the first expected track that it will be 100% when it is not.

File: tests/trackingtest.py (hunk 8 above)
Comment (gferraro) [reply tests/trackingtest.py:31]: I think the case when this is important is if we are testing for false-positives. So if the test recording should have 1 valid track, but it is a noisy track.
When we adjust tracking parameters we may increase the noise and get a lot of false positive tracks. In this case comparing by tracking time is not going to work well.

File: .travis.yml [hunk 12]
```diff
@@ -20,7 +21,12 @@ script:
 - python3 motiontest.py --thermal-config-file smoketest/config.toml -c smoketest/test-config.yaml smoketest/clips/hedgehog.cptv
 - sudo mkdir /var/spool/cptv
 - sudo /home/travis/virtualenv/python3.6.7/bin/python3 piclassify.py --cptv smoketest/clips/hedgehog.cptv  --thermal-config-file smoketest/config.toml -c smoketest/test-config.yaml
-# - python3 classify.py all -c smoketest/test-config.yaml need model  
+- |
```
[the remark this answers is not in the document]
Comment (gferraro) [reply .travis.yml:25]: I am not running these tests on travis, i'm not sure there is any point? but if we are happy with thests it may be worth putting the results for our current tracking somewhere, for easy comparison
